docs/tutorials/debugging.md

```diff
@@ -1,9 +1,145 @@
+
+```{code-cell}
+import jax
+import jax.numpy as jnp
+
+@jax.jit
+def f(x):
+    jax.debug.print("This is `jax.debug.print` of x {x}", x=x)
+    y = jnp.sin(x)
+    jax.debug.print("This is `jax.debug.print` of y {y} 🤯", y=y)
+    return y
+
+f(2.)
+```
+
+{func}`jax.debug.print` can reveal the information about how computations are evaluated.
+
+Here's an example with {func}`jax.vmap`:
+
+```{code-cell}
+def f(x):
+    jax.debug.print("This is `jax.debug.print` of x: {}", x)
+    y = jnp.sin(x)
+    jax.debug.print("This is `jax.debug.print` of y: {}", y)
+    return y
+
+xs = jnp.arange(3.)
+
+jax.vmap(f)(xs)
+```
+
+Here's an example with {func}`jax.lax.map`:
+
+```{code-cell}
+jax.lax.map(f, xs)
+```
+
+Notice the order is different, as {func}`jax.vmap` and {func}`jax.lax.map` compute the same results in different ways. When debugging, the evaluation order details are exactly what you may need to inspect.
+
+Below is an example with {func}`jax.grad`, where {func}`jax.debug.print` only prints the forward pass. In this case, the behavior is similar to Python's `print`, but it's consistent if you apply {func}`jax.jit` during the call.
+
+```{code-cell}
+def f(x):
+    jax.debug.print("This is `jax.debug.print` of x: {}", x)
+    return x ** 2
+
+jax.grad(f)(1.)
+```
+
+Sometimes, when the arguments don't depend on one another, calls to {func}`jax.debug.print` may print them in a different order when staged out with a JAX transformation. If you need the original order, such as `x: ...` first and then `y: ...` second, add the `ordered=True` parameter.
+
+For example:
+
+```{code-cell}
+@jax.jit
+def f(x, y):
+    jax.debug.print("This is `jax.debug.print of x: {}", x, ordered=True)
+    jax.debug.print("This is `jax.debug.print of y: {}", y, ordered=True)
+    return x + y
+```
+
+To learn more about {func}`jax.debug.print` and its Sharp Bits, refer to {ref}`advanced-debugging`.
+
+
+## JAX `debug.breakpoint` for `pdb`-like debugging
+
+**TL;DR** Use {func}`jax.debug.breakpoint` to pause the execution of your JAX program to inspect values.
+
+To pause your compiled JAX program during certain points during debugging, you can use {func}`jax.debug.breakpoint`. The prompt is similar to Python `pdb`, and it allows you to inspect the values in the call stack. In fact, {func}`jax.debug.breakpoint` is an application of {func}`jax.debug.callback` that captures information about the call stack.
+
+To print all available commands during a `breakpoint` debugging session, use the `help` command. (Full debugger commands, the Sharp Bits, its strengths and limitations are covered in {ref}`advanced-debugging`.)
+
+Example:
+
+```{code-cell}
+def breakpoint_if_nonfinite(x):
+  is_finite = jnp.isfinite(x).all()
+  def true_fn(x):
+    pass
+  def false_fn(x):
+    jax.debug.breakpoint()
+  lax.cond(is_finite, true_fn, false_fn, x)
+
+@jax.jit
+def f(x, y):
+  z = x / y
+  breakpoint_if_nonfinite(z)
+  return z
+f(2., 0.) # ==> Pauses during execution
```

Comment:
I suspect we'll need the `skip-execution` tag here, or else the doc build will timeout

Reply:
:+1: